liquidctl/driver/coolit_platinum.py

```diff
@@ -0,0 +1,169 @@
+from liquidctl.pmbus import compute_pec
+from liquidctl.util import clamp
+
+
+LOGGER = logging.getLogger(__name__)
+
+_REPORT_LENGTH = 64
+_WRITE_PREFIX = 0x3F
+_TRAILER_LENGTH = 1
+
+_FEATURE_COOLING = 0x0
+_CMD_GET_STATUS = 0xFF
+_CMD_SET_COOLING = 0x14
+
+_FEATURE_LIGHTING = None
+_CMD_SET_LIGHTING1 = 0b100
+_CMD_SET_LIGHTING2 = 0b100
+
+_SET_COOLING_DATA_OFFSET = 3
+_SET_COOLING_DATA_LENGTH = _REPORT_LENGTH - _SET_COOLING_DATA_OFFSET - _TRAILER_LENGTH
+_FAN1_DATA_OFFSET = 0xB - _SET_COOLING_DATA_OFFSET
+_FAN2_DATA_OFFSET = 0x11 - _SET_COOLING_DATA_OFFSET
+_PUMP_DATA_OFFSET = 0x17 - _SET_COOLING_DATA_OFFSET
+
+# TODO replace with enums and add modes
+_FAN_MODE_FIXED_DUTY = 0x2
+_PUMP_MODE_BALANCED = 0x1
+
+
+def sequence(storage):
+    """Return a generator that produces valid protocol sequence numbers.
+
+    Unstable API.
+
+    Sequence numbers increment across successful invocations of liquidctl, but
+    are not atomic.  The sequence is: 1, 2, 3... 29, 30, 31, 1, 2, 3...
+
+    In the protocol the sequence number is usually shifted left by 3 bits, and
+    a shifted sequence will look like: 8, 16, 24... 232, 240, 248, 8, 16, 24...
+    """
+    while True:
+        seq = storage.load_int('sequence', default=0) % 31 + 1
+        storage.store_int('sequence', seq)
+        yield seq
+
+
+class CoolitPlatinumDriver(UsbHidDriver):
+    """liquidctl driver for Corsair Platinum and PRO XT coolers."""
+
+    SUPPORTED_DEVICES = [
+        # (0x1B1C, ??, None, 'Corsair H100i Platinum SE (experimental)', {}),
+        (0x1B1C, 0x0C18, None, 'Corsair H100i Platinum (experimental)', {}),
+        (0x1B1C, 0x0C17, None, 'Corsair H115i Platinum (experimental)', {}),
+        (0x1B1C, 0x0C20, None, 'Corsair H100i PRO XT (experimental)', {}),
+        (0x1B1C, 0x0C21, None, 'Corsair H115i PRO XT (experimental)', {}),
+        # Note: unknown adjustments needed to support three fans (H150i PRO XT)
+    ]
+
+    def __init__(self, device, description, **kwargs):
+        super().__init__(device, description, **kwargs)
+        # the following fields are only initialized in connect()
+        self._data = None
+        self._sequence = None
+
+    def connect(self, **kwargs):
+        super().connect(**kwargs)
+        ids = '{:04x}_{:04x}'.format(self.vendor_id, self.product_id)
+        # FIXME uniquely identify specific units of the same model
+        self._data = RuntimeStorage(key_prefixes=[ids])
+        self._sequence = sequence(self._data)
+
+    def initialize(self, **kwargs):
+        """Initialize the device."""
+        pass
+
+    def get_status(self, **kwargs):
+        """Get a status report.
+
+        Returns a list of `(property, value, unit)` tuples.
+        """
+        msg = self._send_command(_FEATURE_COOLING, _CMD_GET_STATUS)
+        return [
+            ('Liquid temperature', msg[8] + msg[7] / 255, '°C'),
+            ('Fan 1 speed', int.from_bytes(msg[15:18], byteorder='little'), 'rpm'),
```

Comment:
`msg[15:18]` should be `msg[15:17]`. Reading too many bytes will produce overly large readings.

Reply:
Ops : )

Thanks!